File: 20-valid-parentheses/valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    if (s.length === 0) return true
    if (s.length % 2 !== 0) return false
    
    const stack = []

    for (const char of s) {
        if (char === '(' || char === '{' || char === '[') {
            stack.push(char)
        } else {
            console.log(stack)
            const top = stack.pop()
            if (char === ')' && top !== '(') return false
            if (char === '}' && top !== '{') return false
            if (char === ']' && top !== '[') return false
        }
    }

    return stack.length === 0
};

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        mapping = { '{': '}', '[': ']', '(': ')'}
        stack = []

        for ch in s:
            if ch in mapping:
                stack.append(ch)

            else:
                if not stack or mapping[stack.pop()] != ch:
                    return False

        return not stack